internal/storage/multi_store.go

```diff
@@ -0,0 +1,333 @@
+		logrus.Debugf("Containers for storage drivers %s", k)
+		c, e := v.GetStore().Containers()
+		if e != nil {
+			logrus.Errorf("Getting container for storage driver %s: %v", k, err)
+			err = e
+		}
+		containers = append(containers, c...)
+	}
+
+	return
+}
+
+func (s *multiStoreServer) GraphRoot() string {
+	return s.multiStore.GraphRoot()
+}
+
+func (s *multiStoreServer) ResolveNames(systemContext *ctypes.SystemContext, imageName string) ([]string, error) {
+	logrus.Debugf("ResolveNames for image %s", imageName)
+	for _, s := range s.store {
+		names, err := s.ResolveNames(systemContext, imageName)
+		if err != nil {
+			if err == ErrCannotParseImageID || err == ErrImageMultiplyTagged {
+				return []string{}, err
+			}
+			continue
+		}
+		return names, nil
+	}
+	return []string{}, fmt.Errorf("failed resolving image name for %s not found", imageName)
+}
+
+func (s *multiStoreServer) ImageStatus(systemContext *ctypes.SystemContext, filter string) (*ImageResult, error) {
+	for _, s := range s.store {
+		status, err := s.ImageStatus(systemContext, filter)
+		if err != nil {
+			if err == ErrCannotParseImageID || err == ErrImageMultiplyTagged {
+				return nil, err
+			}
+			continue
+		}
+		return status, nil
+	}
+	return nil, cstorage.ErrImageUnknown
+}
+
+type multiStore struct {
+	defaultStorage string
+	store          map[string]cstorage.Store
+	runRoot        string
+	graphRoot      string
+}
+
+func NewMultiStore(store map[string]cstorage.Store, defaultStorage, runRoot, graphRoot string) MultiStore {
+	return &multiStore{
+		defaultStorage: defaultStorage,
+		store:          store,
+		runRoot:        runRoot,
+		graphRoot:      graphRoot,
+	}
+}
+
+type MultiStore interface {
+	GetDefaultStorageDriver() string
+	GraphDriverName() []string
+	GraphOptions() map[string][]string
+	Containers() ([]cstorage.Container, error)
+	Unmount(id string, force bool) (bool, error)
+	Metadata(id string) (string, error)
+	DeleteContainer(id string) error
+	DeleteImage(id string, commit bool) (layers []string, err error)
+	GetStore() map[string]cstorage.Store
+	RunRoot() string
+	GraphRoot() string
+	GetStoreForContainer(idOrName string) (cstorage.Store, error)
+	GetDefaultStorage() cstorage.Store
+}
+
+func (s *multiStore) GetDefaultStorageDriver() string {
+	return s.defaultStorage
+}
+
+func (s *multiStore) GetDefaultStorage() cstorage.Store {
+	return s.store[s.defaultStorage]
+}
+
+func (s *multiStore) GraphDriverName() (drivers []string) {
+	for k := range s.store {
+		drivers = append(drivers, k)
+	}
+	return
+}
+
+func (s *multiStore) GraphOptions() map[string][]string {
+	graphOptions := make(map[string][]string)
+	for k, v := range s.store {
+		graphOptions[k] = v.GraphOptions()
+	}
+	return graphOptions
+}
+
+func (s *multiStore) Containers() (containers []cstorage.Container, lastError error) {
+	for _, v := range s.store {
+		c, err := v.Containers()
+		// TODO afrosi handles multiple errors
+		if err != nil {
+			lastError = err
+		}
```

Comment:
We can just wrap the errors with errors.Wrap or `fmt.Errorf("my error: %w", err)`